src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -752,26 +832,19 @@ void BaseCouplingScheme::updateTimeAndIterations(
       _totalIterationsCoarseOptimization++;
     }
   } else {
-
     _totalIterationsCoarseOptimization++;
     if (not manifoldmapping)
       _totalIterations++;
 
     _iterationsCoarseOptimization = 1;
     _iterations                   = manifoldmapping ? 0 : 1;
   }
+  _hasDataBeenExchanged = true;
```

Comment:
This is weird. Data does not get exchanged in this function. In general `_hasDataBeenExchanged` is a bit all over the place.

Reply:
I also find the usage of `_hasDataBeenExchanged` mostly confusing. 

At this place we don't need to set it, since it was already set `true` by `exchangeDataAndAccelerate`. There is one exception: If we do serial coupling and we are in the last iteration, we do not set `_hasDataBeenExchanged = true` in `exchangeDataAndAccelerate`. This sounds reasonable. Therefore, I replaced the line above with an assertion in 013aa5e5.

Generally, we could reconsider how `_hasDataBeenExchanged` is used. It always comes "bundled" with `receiveData`:

```
receiveData(getM2N(), getReceiveData());
setHasDataBeenExchanged(true);
```

In my opinion *exchanged* refers to send and receive. We mostly use send and receive in pairs - but not always. We could use `_hasDataBeenSent` and `_hasDataBeenReceived` and finally provide a function
```
bool hasDataBeenExchanged() {
    return _hasDataBeenSent && _hasDataBeenReceived
}
```

Reply:
What do we need `_hasDataBeenSent` for? I guess that `_hasDataBeenExchanged` is actually a `_hasDataBeenReceived`. Needed for `isReadDataAvailable`. 
Or is there also a use case for storing that data has been sent?

Reply:
> I guess that `_hasDataBeenExchanged` is actually a `_hasDataBeenReceived`.

I can check this and then I would like to rename it, since this makes things more clear. This would also allow us to put everything related to setting this bool inside of the receive function. Then one cannot forget setting it, if receive is called (which is a nice side effect).

Reply:
I renamed the member in 0efd0cb. 

I decided to **not** move `checkDataHasBeenReceived()` inside `receiveData()`, since `receiveData()` has to be called multiple times for the `MultiCouplingScheme`. Therefore, we could not use the `PRECICE_ASSERT` that makes sure that `checkDataHasBeenReceived()` is only called once, if we directly include the call in `receiveData()`. I think that the `PRECICE_ASSERT` is nice and useful and therefore don't want to "sacrifice" it.